toolchain/parse/handle_import_and_package.cpp

```diff
@@ -0,0 +1,155 @@
+static auto ExitOnParseError(Context& context, Context::StateStackEntry state,
+                             NodeKind directive) {
+  auto semi_token = context.SkipPastLikelyEnd(state.token);
+  return context.AddNode(directive, semi_token ? *semi_token : state.token,
+                         state.subtree_start,
+                         /*has_error=*/true);
+}
+
+// Handles the main parsing of `import`/`package`. It's expected that the
+// introducer is already added.
+static auto HandleImportAndPackage(Context& context,
+                                   Context::StateStackEntry state,
+                                   NodeKind directive, bool expect_api_or_impl)
+    -> void {
+  if (!context.ConsumeAndAddLeafNodeIf(Lex::TokenKind::Identifier,
+                                       NodeKind::Name)) {
+    CARBON_DIAGNOSTIC(ExpectedIdentifierAfterKeyword, Error,
+                      "Expected identifier after `{0}`.", Lex::TokenKind);
+    context.emitter().Emit(*context.position(), ExpectedIdentifierAfterKeyword,
+                           context.tokens().GetKind(state.token));
+    ExitOnParseError(context, state, directive);
+    return;
+  }
+
+  bool library_parsed = false;
+  if (auto library_token = context.ConsumeIf(Lex::TokenKind::Library)) {
+    auto library_start = context.tree().size();
+
+    if (!context.ConsumeAndAddLeafNodeIf(Lex::TokenKind::StringLiteral,
+                                         NodeKind::Literal)) {
+      CARBON_DIAGNOSTIC(
+          ExpectedLibraryName, Error,
+          "Expected a string literal to specify the library name.");
+      context.emitter().Emit(*context.position(), ExpectedLibraryName);
+      ExitOnParseError(context, state, directive);
+      return;
+    }
+
+    context.AddNode(NodeKind::Library, *library_token, library_start,
+                    /*has_error=*/false);
+    library_parsed = true;
+  }
+
+  auto next_kind = context.tokens().GetKind(*(context.position()));
+  if (!library_parsed && next_kind == Lex::TokenKind::StringLiteral) {
+    // If we come acroess a string literal and we didn't parse `library
+    // "..."` yet, then most probably the user forgot to add `library`
+    // before the library name.
+    CARBON_DIAGNOSTIC(MissingLibraryKeyword, Error,
+                      "Missing `library` keyword.");
+    context.emitter().Emit(*context.position(), MissingLibraryKeyword);
+    ExitOnParseError(context, state, directive);
+    return;
+  }
+
+  if (expect_api_or_impl) {
+    switch (next_kind) {
+      case Lex::TokenKind::Api: {
+        context.AddLeafNode(NodeKind::PackageApi, context.Consume());
+        break;
+      }
+      case Lex::TokenKind::Impl: {
+        context.AddLeafNode(NodeKind::PackageImpl, context.Consume());
+        break;
+      }
+      default: {
+        CARBON_DIAGNOSTIC(ExpectedApiOrImpl, Error,
+                          "Expected a `api` or `impl`.");
```

Comment:
```suggestion
                          "Expected an `api` or `impl`.");
```
... or maybe drop the article entirely?
```suggestion
                          "Expected `api` or `impl`.");
```

Reply:
Looks like this PR is just moving this diagnostic around, so I think it's fine to defer this, but if you want to fix it in passing, please do :)